### ISSUE TITLE
@bfan @caozhangjie

### ISSUE THREAD

**Issue by LiShuiYu** [thuml/HashNet#35]
@bfan @caozhangjie 
I add the weight in pytorch version(without c).
```
def pairwise_loss(outputs1,outputs2,label1,label2):
    similarity = Variable(torch.mm(label1.data.float(), label2.data.float().t()) > 0).float()
    dot_product = torch.mm(outputs1, outputs2.t())
    #exp_product = torch.exp(dot_product)

    mask_positive = similarity.data > 0
    mask_negative = similarity.data <= 0
    exp_loss = torch.log(1+torch.exp(-torch.abs(dot_product))) + torch.max(dot_product, Variable(torch.FloatTensor([0.]).cuda()))-similarity * dot_product
    #weight
    S1 = torch.sum(mask_positive.float())
    S0 = torch.sum(mask_negative.float())
    S = S0+S1
    exp_loss[similarity.data > 0] = exp_loss[similarity.data > 0] * (S / S1)
    exp_loss[similarity.data <= 0] = exp_loss[similarity.data <= 0] * (S / S0)

    loss = torch.sum(exp_loss) / S

    #exp_loss = torch.sum(torch.log(1 + exp_product) - similarity * dot_product)

    return loss
```

_Originally posted by @soon-will in https://github.com/thuml/HashNet/issues/17#issuecomment-443137529_

**Comment by LiShuiYu** [comment]
Excuse me, I don't understand the principle of this supplementary pytorch version loss function? Can someone explain this?
